File: backend/app/services/typing_engine.py
from typing import List, Dict, Optional, Tuple
from dataclasses import dataclass, field
import time
from collections import defaultdict


@dataclass
class TypingMetrics:
    gross_wpm: float = 0.0
    net_wpm: float = 0.0
    accuracy: float = 0.0
    error_percentage: float = 0.0
    key_depression_count: int = 0
    backspace_count: int = 0
    pause_count: int = 0
    total_pause_duration: float = 0.0
    avg_pause_duration: float = 0.0
    longest_pause_duration: float = 0.0
    time_utilization_percentage: float = 0.0
    consistency_score: float = 0.0
    typing_rhythm_score: float = 0.0
    error_zones: Dict = field(default_factory=dict)
    weak_words: List[str] = field(default_factory=list)
    common_mistypes: Dict = field(default_factory=dict)
    left_hand_errors: int = 0
    right_hand_errors: int = 0
    shift_key_errors: int = 0
    number_row_errors: int = 0


class TypingEngine:
    PAUSE_THRESHOLD_MS = 2000
    LEFT_HAND_KEYS = set("qwertasdfgzxcvb")
    RIGHT_HAND_KEYS = set("yuiophjklnm")
    NUMBER_ROW_KEYS = set("1234567890")
    SHIFT_KEYS = set("~!@#$%^&*()_+{}|:\"<>?QWERTYUIOPASDFGHJKLZXCVBNM")

    def __init__(self):
        self._sessions: Dict[str, Dict] = {}

    def create_session(self, test_id: str, original_content: str, duration_seconds: int):
        self._sessions[test_id] = {
            "original": original_content,
            "duration": duration_seconds,
            "events": [],
            "start_time": time.time(),
            "pauses": [],
            "backspaces": 0,
            "total_chars": 0,
            "error_count": 0,
            "corrections": 0,
            "char_timings": [],
            "last_event_time": None,
        }
        return test_id

    def record_event(self, test_id: str, key: str, timestamp_ms: int, duration_ms: int = 0):
        session = self._sessions.get(test_id)
        if not session:
            return

        current_time = time.time() * 1000

        if session["last_event_time"] is not None:
            gap = current_time - session["last_event_time"]
            if gap > self.PAUSE_THRESHOLD_MS:
                session["pauses"].append({
                    "start_ms": int(session["last_event_time"]),
                    "end_ms": int(current_time),
                    "duration_ms": int(gap),
                })

        session["last_event_time"] = current_time

        is_backspace = key in ("Backspace", "Delete")
        if is_backspace:
            session["backspaces"] += 1

        session["events"].append({
            "key": key,
            "timestamp_ms": timestamp_ms,
            "duration_ms": duration_ms,
            "is_backspace": is_backspace,
        })
        session["char_timings"].append({
            "key": key,
            "time": timestamp_ms,
            "duration": duration_ms,
        })

    def compute_metrics(self, test_id: str, typed_content: str, original_content: str, time_taken_seconds: float) -> TypingMetrics:
        session = self._sessions.get(test_id)
        if not session:
            return TypingMetrics()

        metrics = TypingMetrics()
        events = session["events"]
        char_timings = session["char_timings"]
        pauses = session["pauses"]

        metrics.key_depression_count = len(typed_content)
        metrics.backspace_count = session["backspaces"]
        metrics.pause_count = len(pauses)

        if pauses:
            metrics.total_pause_duration = sum(p["duration_ms"] for p in pauses) / 1000.0
            metrics.avg_pause_duration = metrics.total_pause_duration / len(pauses)
            metrics.longest_pause_duration = max(p["duration_ms"] for p in pauses) / 1000.0

        total_duration_ms = time_taken_seconds * 1000
        if total_duration_ms > 0:
            active_time = max(0, total_duration_ms - (metrics.total_pause_duration * 1000))
            metrics.time_utilization_percentage = round((active_time / total_duration_ms) * 100, 2)

        metrics.consistency_score = self._calculate_consistency(char_timings)
        metrics.typing_rhythm_score = self._calculate_rhythm(char_timings)

        metrics.left_hand_errors, metrics.right_hand_errors, metrics.shift_key_errors, metrics.number_row_errors = self._analyze_hand_errors(events, original_content)

        metrics.weak_words = self._identify_weak_words(typed_content, original_content)
        metrics.common_mistypes = self._find_common_mistypes(events, original_content)
        metrics.error_zones = self._generate_error_zones(events, original_content)

        del self._sessions[test_id]
        return metrics

    def _calculate_consistency(self, char_timings: List[Dict]) -> float:
        if len(char_timings) < 10:
            return 100.0

        intervals = []
        for i in range(1, len(char_timings)):
            gap = char_timings[i]["time"] - char_timings[i - 1]["time"]
            if 0 < gap < 5000:
                intervals.append(gap)

        if not intervals:
            return 0.0

        avg_interval = sum(intervals) / len(intervals)
        variance = sum((i - avg_interval) ** 2 for i in intervals) / len(intervals)
        std_dev = variance ** 0.5

        cv = std_dev / avg_interval if avg_interval > 0 else 1.0
        consistency = max(0, min(100, 100 - (cv * 50)))
        return round(consistency, 2)

    def _calculate_rhythm(self, char_timings: List[Dict]) -> float:
        if len(char_timings) < 20:
            return 100.0

        bigrams = []
        for i in range(1, len(char_timings)):
            gap = char_timings[i]["time"] - char_timings[i - 1]["time"]
            if 0 < gap < 3000:
                bigrams.append(gap)

        if not bigrams:
            return 0.0

        smooth = sum(bigrams) / len(bigrams)
        rhythm_penalties = sum(1 for b in bigrams if abs(b - smooth) > smooth * 1.5)
        penalty_ratio = rhythm_penalties / len(bigrams)
        score = max(0, min(100, 100 - (penalty_ratio * 100)))
        return round(score, 2)

    def _analyze_hand_errors(self, events: List[Dict], original: str) -> Tuple[int, int, int, int]:
        left_errors = 0
        right_errors = 0
        shift_errors = 0
        number_errors = 0

        for event in events:
            key = event.get("key", "")
            if key in self.LEFT_HAND_KEYS:
                left_errors += 1
            elif key in self.RIGHT_HAND_KEYS:
                right_errors += 1
            elif key in self.SHIFT_KEYS:
                shift_errors += 1
            elif key in self.NUMBER_ROW_KEYS:
                number_errors += 1

        return left_errors, right_errors, shift_errors, number_errors

    def _identify_weak_words(self, typed: str, original: str) -> List[str]:
        typed_words = typed.split()
        original_words = original.split()
        weak_words = []

        import Levenshtein
        for i in range(min(len(typed_words), len(original_words))):
            if typed_words[i] != original_words[i]:
                weak_words.append(original_words[i])

        word_freq = defaultdict(int)
        for w in weak_words:
            word_freq[w] += 1

        sorted_weak = sorted(word_freq.items(), key=lambda x: -x[1])
        return [w for w, _ in sorted_weak[:20]]

    def _find_common_mistypes(self, events: List[Dict], original: str) -> Dict:
        mistypes = defaultdict(int)
        for event in events:
            key = event.get("key", "")
            if event.get("is_error"):
                if key and len(key) == 1:
                    mistypes[key] += 1
        return dict(sorted(mistypes.items(), key=lambda x: -x[1])[:10])

    def _generate_error_zones(self, events: List[Dict], original: str) -> Dict:
        return {
            "left_hand": {"errors": 0, "keys": list(self.LEFT_HAND_KEYS)},
            "right_hand": {"errors": 0, "keys": list(self.RIGHT_HAND_KEYS)},
            "number_row": {"errors": 0, "keys": list(self.NUMBER_ROW_KEYS)},
            "shift_key": {"errors": 0},
        }


typing_engine = TypingEngine()

File: frontend/src/server/services/typing-engine.ts
import { errorEngine } from './error-engine';

export interface TypingMetrics {
  grossWpm: number;
  netWpm: number;
  accuracy: number;
  errorPercentage: number;
  keyDepressionCount: number;
  backspaceCount: number;
  pauseCount: number;
  totalPauseDuration: number;
  avgPauseDuration: number;
  longestPauseDuration: number;
  timeUtilizationPercentage: number;
  consistencyScore: number;
  typingRhythmScore: number;
  errorZones: Record<string, { errors: number; keys: string[] }>;
  weakWords: string[];
  commonMistypes: Record<string, number>;
  leftHandErrors: number;
  rightHandErrors: number;
  shiftKeyErrors: number;
  numberRowErrors: number;
}

interface CharTiming {
  key: string;
  time: number;
  duration: number;
}

interface PauseInfo {
  startMs: number;
  endMs: number;
  durationMs: number;
}

interface KeystrokeEvent {
  key: string;
  timestampMs: number;
  durationMs: number;
  isBackspace?: boolean;
  isError?: boolean;
}

interface SessionData {
  original: string;
  duration: number;
  events: KeystrokeEvent[];
  startTime: number;
  pauses: PauseInfo[];
  backspaces: number;
  totalChars: number;
  errorCount: number;
  corrections: number;
  charTimings: CharTiming[];
  lastEventTime: number | null;
}

export class TypingEngine {
  private static readonly PAUSE_THRESHOLD_MS = 2000;
  private static readonly LEFT_HAND_KEYS = new Set('qwertasdfgzxcvb');
  private static readonly RIGHT_HAND_KEYS = new Set('yuiophjklnm');
  private static readonly NUMBER_ROW_KEYS = new Set('1234567890');
  private static readonly SHIFT_KEYS = new Set('~!@#$%^&*()_+{}|:"<>?QWERTYUIOPASDFGHJKLZXCVBNM');
  private static readonly SESSION_TTL_MS = 30 * 60 * 1000;

  private sessions: Map<string, SessionData> = new Map();
  private cleanupTimer: ReturnType<typeof setInterval> | null = null;

  constructor() {
    this.cleanupTimer = setInterval(() => this.evictStale(), 60_000);
    this.cleanupTimer.unref();
  }

  dispose(): void {
    if (this.cleanupTimer) {
      clearInterval(this.cleanupTimer);
      this.cleanupTimer = null;
    }
    this.sessions.clear();
  }

  private evictStale(): void {
    const now = Date.now();
    for (const [id, session] of this.sessions) {
      if (now - session.startTime > TypingEngine.SESSION_TTL_MS) {
        this.sessions.delete(id);
      }
    }
  }

  createSession(testId: string, originalContent: string, durationSeconds: number): string {
    this.sessions.set(testId, {
      original: originalContent,
      duration: durationSeconds,
      events: [],
      startTime: Date.now(),
      pauses: [],
      backspaces: 0,
      totalChars: 0,
      errorCount: 0,
      corrections: 0,
      charTimings: [],
      lastEventTime: null,
    });
    return testId;
  }

  recordEvent(testId: string, key: string, timestampMs: number, durationMs: number = 0): void {
    const session = this.sessions.get(testId);
    if (!session) return;

    const currentTime = Date.now();

    if (session.lastEventTime !== null) {
      const gap = currentTime - session.lastEventTime;
      if (gap > TypingEngine.PAUSE_THRESHOLD_MS) {
        session.pauses.push({
          startMs: Math.round(session.lastEventTime),
          endMs: Math.round(currentTime),
          durationMs: Math.round(gap),
        });
      }
    }

    session.lastEventTime = currentTime;

    const isBackspace = key === 'Backspace' || key === 'Delete';
    if (isBackspace) {
      session.backspaces += 1;
    }

    session.events.push({ key, timestampMs, durationMs, isBackspace });
    session.charTimings.push({ key, time: timestampMs, duration: durationMs });
  }

  computeMetrics(
    testId: string,
    typedContent: string,
    originalContent: string,
    timeTakenSeconds: number,
  ): TypingMetrics {
    const session = this.sessions.get(testId);
    if (!session) {
      return this.emptyMetrics();
    }

    const events = session.events;
    const charTimings = session.charTimings;
    const pauses = session.pauses;

    const keyDepressionCount = typedContent.length;
    const backspaceCount = session.backspaces;
    const pauseCount = pauses.length;

    let totalPauseDuration = 0;
    let avgPauseDuration = 0;
    let longestPauseDuration = 0;

    if (pauses.length > 0) {
      totalPauseDuration = pauses.reduce((sum, p) => sum + p.durationMs, 0) / 1000;
      avgPauseDuration = totalPauseDuration / pauses.length;
      longestPauseDuration = Math.max(...pauses.map(p => p.durationMs)) / 1000;
    }

    const totalDurationMs = timeTakenSeconds * 1000;
    let timeUtilizationPercentage = 0;
    if (totalDurationMs > 0) {
      const activeTime = Math.max(0, totalDurationMs - (totalPauseDuration * 1000));
      timeUtilizationPercentage = Math.round((activeTime / totalDurationMs) * 10000) / 100;
    }

    const consistencyScore = this.calculateConsistency(charTimings);
    const typingRhythmScore = this.calculateRhythm(charTimings);

    const [leftHandErrors, rightHandErrors, shiftKeyErrors, numberRowErrors] =
      this.analyzeHandErrors(events, originalContent);

    const weakWords = this.identifyWeakWords(typedContent, originalContent);
    const commonMistypes = this.findCommonMistypes(events, originalContent);
    const errorZones = this.generateErrorZones(events, originalContent);

    this.sessions.delete(testId);

    const report = errorEngine.evaluate(originalContent, typedContent, timeTakenSeconds);

    return {
      grossWpm: report.grossWpm,
      netWpm: report.netWpm,
      accuracy: report.accuracy,
      errorPercentage: report.errorPercentage,
      keyDepressionCount,
      backspaceCount,
      pauseCount,
      totalPauseDuration: Math.round(totalPauseDuration * 100) / 100,
      avgPauseDuration: Math.round(avgPauseDuration * 100) / 100,
      longestPauseDuration: Math.round(longestPauseDuration * 100) / 100,
      timeUtilizationPercentage,
      consistencyScore,
      typingRhythmScore,
      errorZones,
      weakWords,
      commonMistypes,
      leftHandErrors,
      rightHandErrors,
      shiftKeyErrors,
      numberRowErrors,
    };
  }

  private calculateConsistency(charTimings: CharTiming[]): number {
    if (charTimings.length < 10) return 100;

    const intervals: number[] = [];
    for (let i = 1; i < charTimings.length; i++) {
      const gap = charTimings[i].time - charTimings[i - 1].time;
      if (gap > 0 && gap < 5000) {
        intervals.push(gap);
      }
    }

    if (intervals.length === 0) return 0;

    const avgInterval = intervals.reduce((a, b) => a + b, 0) / intervals.length;
    const variance = intervals.reduce((sum, i) => sum + (i - avgInterval) ** 2, 0) / intervals.length;
    const stdDev = Math.sqrt(variance);
    const cv = avgInterval > 0 ? stdDev / avgInterval : 1;
    const consistency = Math.max(0, Math.min(100, 100 - cv * 50));
    return Math.round(consistency * 100) / 100;
  }

  private calculateRhythm(charTimings: CharTiming[]): number {
    if (charTimings.length < 20) return 100;

    const bigrams: number[] = [];
    for (let i = 1; i < charTimings.length; i++) {
      const gap = charTimings[i].time - charTimings[i - 1].time;
      if (gap > 0 && gap < 3000) {
        bigrams.push(gap);
      }
    }

    if (bigrams.length === 0) return 0;

    const smooth = bigrams.reduce((a, b) => a + b, 0) / bigrams.length;
    const rhythmPenalties = bigrams.filter(b => Math.abs(b - smooth) > smooth * 1.5).length;
    const penaltyRatio = rhythmPenalties / bigrams.length;
    const score = Math.max(0, Math.min(100, 100 - penaltyRatio * 100));
    return Math.round(score * 100) / 100;
  }

  private analyzeHandErrors(
    events: KeystrokeEvent[],
    _original: string,
  ): [number, number, number, number] {
    let leftErrors = 0, rightErrors = 0, shiftErrors = 0, numberErrors = 0;

    for (const event of events) {
      const key = event.key;
      if (TypingEngine.LEFT_HAND_KEYS.has(key)) leftErrors++;
      else if (TypingEngine.RIGHT_HAND_KEYS.has(key)) rightErrors++;
      else if (TypingEngine.SHIFT_KEYS.has(key)) shiftErrors++;
      else if (TypingEngine.NUMBER_ROW_KEYS.has(key)) numberErrors++;
    }

    return [leftErrors, rightErrors, shiftErrors, numberErrors];
  }

  private identifyWeakWords(typed: string, original: string): string[] {
    const typedWords = typed.split(/\s+/);
    const originalWords = original.split(/\s+/);
    const weakWords: string[] = [];

    for (let i = 0; i < Math.min(typedWords.length, originalWords.length); i++) {
      if (typedWords[i] !== originalWords[i]) {
        weakWords.push(originalWords[i]);
      }
    }

    const wordFreq = new Map<string, number>();
    for (const w of weakWords) {
      wordFreq.set(w, (wordFreq.get(w) || 0) + 1);
    }

    return [...wordFreq.entries()]
      .sort((a, b) => b[1] - a[1])
      .slice(0, 20)
      .map(([w]) => w);
  }

  private findCommonMistypes(events: KeystrokeEvent[], _original: string): Record<string, number> {
    const mistypes = new Map<string, number>();
    for (const event of events) {
      if (event.isError) {
        if (event.key && event.key.length === 1) {
          mistypes.set(event.key, (mistypes.get(event.key) || 0) + 1);
        }
      }
    }
    const sorted = [...mistypes.entries()].sort((a, b) => b[1] - a[1]).slice(0, 10);
    return Object.fromEntries(sorted);
  }

  private generateErrorZones(
    _events: KeystrokeEvent[],
    _original: string,
  ): Record<string, { errors: number; keys: string[] }> {
    return {
      left_hand: { errors: 0, keys: [...TypingEngine.LEFT_HAND_KEYS] },
      right_hand: { errors: 0, keys: [...TypingEngine.RIGHT_HAND_KEYS] },
      number_row: { errors: 0, keys: [...TypingEngine.NUMBER_ROW_KEYS] },
      shift_key: { errors: 0, keys: [] },
    };
  }

  private emptyMetrics(): TypingMetrics {
    return {
      grossWpm: 0, netWpm: 0, accuracy: 0, errorPercentage: 0,
      keyDepressionCount: 0, backspaceCount: 0, pauseCount: 0,
      totalPauseDuration: 0, avgPauseDuration: 0, longestPauseDuration: 0,
      timeUtilizationPercentage: 0, consistencyScore: 0, typingRhythmScore: 0,
      errorZones: {}, weakWords: [], commonMistypes: {},
      leftHandErrors: 0, rightHandErrors: 0, shiftKeyErrors: 0, numberRowErrors: 0,
    };
  }
}

export const typingEngine = new TypingEngine();
